### DISCUSSION TITLE
Enhancement: Perplexity Completion response incl. citations

### DISCUSSION THREAD

**Discussion by djuillard** [danny-avila/LibreChat#4692] [Feature Requests & Suggestions]
### What features would you like to see added?

Las week, Perplexity announced that citations (= url sources) are by default returned with chat completions. This is included into a citations object within the completion output response. I would like to be able to display the citations directly in the chat window, like it is in the perplexity web app. 

### More details

Exemple of API query towards perplexity : 

{
    "id": "5db80598-305a-473e-98ba-eb9dbca8a62d",
    "model": "llama-3.1-sonar-small-128k-online",
    "created": 1731340613,
    "usage": {
        "prompt_tokens": 7,
        "completion_tokens": 497,
        "total_tokens": 504
    },
    "citations": [
        "https://www.theatlantic.com/family/archive/2021/10/meaning-life-macronutrients-purpose-search/620440/",
        "https://en.wikipedia.org/wiki/Meaning_of_life",
        "https://greatergood.berkeley.edu/article/item/three_ways_to_see_meaning_in_your_life",
        "https://philosophynow.org/issues/147/Why_Youre_Probably_Wrong_About_The_Meaning_of_Life",
        "https://iankoniak.com/what-is-the-meaning-of-life/"
    ],
    "object": "chat.completion",
    "choices": [
        {
            "index": 0,
            "finish_reason": "stop",
            "message": {
                "role": "assistant",
                "content": "The meaning of life is a deeply personal and complex question that has been explored by philosophers, theologians, and psychologists across various cultures and ideologies. There is no consensus on a definitive answer, and the search for meaning continues to be a significant aspect of human existence.\n\n### Key Dimensions of Meaning\n\n1. **Coherence**: This dimension involves the understanding that events in life fit together for a reason. It is the belief that things happen in your life for a purpose, even if you may not fully comprehend it at the moment[1][3].\n\n2. **Purpose**: This refers to the existence of goals and aims. It is the belief that you are alive to do something specific, such as sharing secrets to happiness or spreading love abundantly[1][3].\n\n3. **Significance**: This dimension involves the sense that your life matters. It is the confidence that the world would be poorer if you didn't exist, giving your life inherent value[1][3].\n\n### Philosophical Perspectives\n\n1. **Subjective vs. Objective Meaning**: Some philosophers argue that meaning in life is subjective, derived from personal experiences and goals. Others propose that there are objective criteria for what constitutes meaningful life, such as fulfilling goals and contributing to society[2][4].\n\n2. **Creating vs. Detecting Meaning**: While some argue that meaning must be actively constructed through deep self-reflection or significant life events, research suggests that most people find meaning in their daily lives through their relationships, work, and spiritual beliefs[3][4].\n\n3. **Activities and Experiences**: Philosophers identify various activities and experiences as meaningful, including loving relationships, creativity, personal excellence, moral goodness, altruism, knowledge, transcendence, and achievement[4].\n\n### Practical Approaches\n\n1. **Pursuing Full Potential**: One practical approach is to pursue your full potential and become the best version of yourself in all areas of life. This philosophy emphasizes the joy in the journey rather than just the destination[5].\n\n2. **Cultivating Coherence and Purpose**: By focusing on coherence (making sense of life) and purpose (orienting towards fulfilling goals), individuals can enhance their experience of meaning. This involves detecting the meaning already present in daily life and actively pursuing fulfilling goals[3].\n\nIn summary, the meaning of life is multifaceted and highly individualized. It involves a combination of coherence, purpose, and significance, which can be cultivated through various activities and experiences."
            },
            "delta": {
                "role": "assistant",
                "content": ""
            }
        }
    ]
}

### Pictures

UI and completion processing.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Dombom123** [comment]
@danny-avila Hey! im also super interested in this Enhancement. Any chances that this will be implemented?

**Comment by Shura1oplot** [comment]
That addition would be fantastic!

**Comment by sbogeFS** [comment]
I completeley agree that this addition wouldb e great. Moreover self hosted open ai instances, for instance in azure, allow vector/semantic seach on uploaded data which in turn are automatically evaluated and returned as text summary with citations regarding the content of the files. The file content is then cited with short text, the JSON format is as follows:
```{
  role: "assistant",
  content: "The Nice App is a usefull app for everyone who likes apps.",
  end_turn: true,
  context: {
    citations: [
      {
        content: "Text... cited...for...nice app..part 1.",
        title: "Nice App documentation",
        url: "/documentation.pdf",
        filepath: "./res/documentation.pdf",
        chunk_id: "0",
      },
            {
        content: "Text... cited...for...nice app..part 2.",
        title: "Nice App User Guide",
        url: "/guide.pdf",
        filepath: "./res/guide.pdf",
        chunk_id: "1",
      },
    ],
    intent: "[\"What ist the nice app?\"]",
  },
}```

**Reply by sbogeFS** [reply]
Also mentioned in #1207 regarding selfhosted azure instances and in #734 regarding usage of plugins. In context of the latter, I wonder where I can find documentation with respect to the topic of plugins and if these are supported in the future? Is there da new update on that topic?

**Comment by larryb821** [comment]
would love to see the citations appear in librechat UI.  +1 will buy coffee for devs! 

![image](https://github.com/user-attachments/assets/79c07c39-84de-47f9-8064-72a9579bf83d)

**Reply by rubentalstra** [reply]
Sounds good 😌 I will work next on this then. 🤓

**Reply by ion-elgreco** [reply]
@rubentalstra curious to know the OpenAI assistants api returns citations in their response, is this currently used by any chance?

**Reply by sbogeFS** [reply]
@ion-elgreco OpenAi does and it would be perfectly fine if this would be reconginzed.

**Comment by Varial17** [comment]
anyone know if we can do this?

**Comment by anphex** [comment]
I am also very interested in this citations feature since Perplexity gets a lot of love lately. Sonar Pro and Sonar Reasoning Pro feel really good. but I can't see the links. Also there is currently really no cheap chat title summary model in the current Perplexity API, would love to just use gpt4o-mini for all of my endpoints, even the custom ones.

**Comment by aleibovici** [comment]
One more vote for any support for citations with the complexity endpoint. Thanks!

**Comment by Dombom123** [comment]
I'm still waiting for feedback from @danny-avila but in the meantime im using this workaround:

Put something like this into the System message for perplexity endpoints:

```
Additional Rules:
- Always print the full URL instead of just the Citation Number
- Include a List of References with the full URL in your answer
- Always cite the sources of your web search
```

**Reply by aleibovici** [reply]
Thanks @Dombom123 .. works like a charm!

**Reply by larryb821** [reply]
> Put something like this into the System message for perplexity endpoints:

Hi @Dombom123 can you share the yaml for system message? I tried with the code below and i dont think its working. TIA

custom:
    # Preplexity AI Example
    - name: "Perplexity"
      apiKey: "${PERPLEXITY_API_KEY}"
      baseURL: "https://api.perplexity.ai/"
      models:
        default: [
          "sonar-reasoning-pro",
          "sonar-reasoning",
          "sonar-pro",
          "sonar"
          ]
        fetch: false # fetching list of models is not supported
      titleConvo: true
      titleModel: "current_model"
      forcePrompt: false
      titleMessageRole: "user"
      modelDisplayLabel: "Perplexity"
      dropParams: ["stop", "user", "frequency_penalty", "presence_penalty"]
      promptPrefix: |
        Additional Rules:
        - Always print the full URL instead of just the Citation Number
        - Include a List of References with the full URL in your answer
        - Always cite the sources of your web search

**Reply by Dombom123** [reply]
im not using librechat.yaml to set my system message. im using the agents endpoint. you can set custom system messages just by using the ui.

**Reply by voelspriet** [reply]
intendation

**Comment by marlonka** [comment]
+1 a fix is really appreciated that shows sources without system message change

**Comment by dannykorpan** [comment]
@larryb821
This workaround is working. I'm using Perplexity over OpenRouter, but I think it's also working via Perplexity API.

librechat.yaml
```
...
endpoints:
  custom:
    - name: "Perplexity"
      apiKey: "..."
      baseURL: "https://openrouter.ai/api/v1"
      models:
        default: ["perplexity/sonar-reasoning-pro","perplexity/sonar-pro","perplexity/sonar-deep-research"]
        fetch: false
      titleConvo: true
      titleModel: "openai/gpt-4.1-nano"
      summarize: false
      forcePrompt: false
      modelDisplayLabel: "Perplexity"
      dropParams:
        - "stop"
modelSpecs:
  enforce: false
  list:
    - name: "perplexity-sonar-pro"
      label: "Perplexity: Sonar Pro"
      preset:
        endpoint: "Perplexity"
        model: "perplexity/sonar-pro"
        temperature: 0.3
        promptPrefix: "Additional Rules: 1. Always print the full URL in brackets after the Cititation Number instead of just the Citation Number; 2. Ensure responses are concise and relevant; 3. Always cite the sources of your web search"
    - name: "perplexity-reasoning-sonar-pro"
      label: "Perplexity: Sonar Reasoning Pro"
      preset:
        endpoint: "Perplexity"
        model: "perplexity/sonar-pro"
        temperature: 0.3
        promptPrefix: "Additional Rules: 1. Always print the full URL in brackets after the Cititation Number instead of just the Citation Number; 2. Ensure responses are concise and relevant; 3. Always cite the sources of your web search"
    - name: "perplexity-deep-research"
      label: "Perplexity: Deep Research"
      preset:
        endpoint: "Perplexity"
        model: "perplexity/sonar-deep-research"
        temperature: 0.3
        promptPrefix: "Additional Rules: 1. Always print the full URL in brackets after the Cititation Number instead of just the Citation Number; 2. Ensure responses are concise and relevant; 3. Always cite the sources of your web search"
 ```

**Comment by guygud** [comment]
+1 for this feature

**Comment by jgdub** [comment]
I would like to see this as well.

**Comment by voelspriet** [comment]
Me too

**Comment by enieuwy** [comment]
Also hope to see this!

**Comment by Ziyann** [comment]
It'd be cool if this could be made somewhat generic, as more or less all frontier models return citation/search metadata now. LiteLLM and OpenRouter also returns it for them (for custom endpoints). The format of the data across them is similar, so it shouldn't be too hard to parse and convert to a unified format for the UI to show them (but we lack the UI support currently).

**Reply by petervcook** [reply]
For reference here is the documentation for how OpenRouter returns citations: https://openrouter.ai/docs/features/web-search#parsing-web-search-results

**Comment by Shura1oplot** [comment]
I've implemented a [hack](https://github.com/Shura1oplot/librechat_litellm_proxy) using LiteLLM to insert web search results in the end of the AI response.
